crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -228,67 +239,120 @@ impl<C: CursorType + Eq + Clone + Send + Sync + 'static> Page<C> {
             results
         };
 
-        // Detect whether the results imply the existence of a previous or next page.
-        let (prev, next, prefix, suffix) = match (
-            self.after(),
-            results.first(),
-            results.last(),
-            self.before(),
-            self.end,
-        ) {
-            // Results came back empty, despite supposedly including the `after` and `before`
-            // cursors, so the bounds must have been invalid, no matter which end the page was
-            // drawn from.
-            (_, None, _, _, _) | (_, _, None, _, _) => {
-                return Ok((false, false, vec![].into_iter()));
-            }
-
-            // Page drawn from the front, and the cursor for the first element does not match
-            // `after`. This implies the bound was invalid, so we return an empty result.
-            (Some(a), Some(f), _, _, End::Front) if f.cursor() != *a => {
-                return Ok((false, false, vec![].into_iter()));
-            }
+        Ok(self.paginate_results(
+            results.first().map(|f| f.cursor()),
+            results.last().map(|l| l.cursor()),
+            results,
+        ))
+    }
 
-            // Similar to above case, but for back of results.
-            (_, _, Some(l), Some(b), End::Back) if l.cursor() != *b => {
-                return Ok((false, false, vec![].into_iter()));
+    pub(crate) fn paginate_raw_query<T>(
+        &self,
+        conn: &mut Conn<'_>,
+        query_fn: impl Fn() -> RawQuery,
```

Comment:
I don't think you need `paginate_raw_query` to accept a query thunk, because `RawQuery` can implement `clone` just fine.